clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,232 @@
+                            LintResult::NothingCaptured | LintResult::Consumed => LintResult::Consumed,
+                            LintResult::NeedMove => LintResult::NeedMove,
+                        }
+                    },
+                }
+            }
+        }
+
+        let note_msg = match lint_result {
+            LintResult::NothingCaptured => "there were no captured variables, so the `move` is unnecessary",
+            LintResult::Consumed => {
+                "there were consumed variables, but no borrowed variables, so the `move` is unnecessary"
+            },
+            LintResult::NeedMove => {
+                // there was a value which would be borrowed if it weren't for the move keyword,
+                // so we should keep it, as removing it would change semantics.
+                return;
+            },
+        };
+
+        span_lint_and_then(
+            cx,
+            NEEDLESS_MOVE,
+            expr.span,
+            "you seem to use `move`, but the `move` is unnecessary",
+            |diag| {
+                diag.suggest_remove_item(cx, move_kw, "remove the `move`", Applicability::MachineApplicable);
+                diag.note(note_msg);
+            },
+        );
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for NeedlessMove {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        if let ExprKind::Closure(closure) = &expr.kind {
+            Self::check_closure(cx, expr, closure);
+        }
+    }
+}
+
+enum LintResult {
+    /// do not remove the `move` keyword.
+    NeedMove,
+    Consumed,
+    NothingCaptured,
+}
+
+#[derive(Debug, Default)]
+struct MovedVariablesCtxt<'tcx> {
+    // for each base variable, we remember:
+    /// The places where it was captured (and consumed, e.g. moved into the closure).
+    moved: Vec<(euv::Place<'tcx>, HirId)>,
+    /// The places where it was captured by reference (and not consumed).
+    captured: Vec<(euv::Place<'tcx>, HirId, ty::BorrowKind)>,
+}
+
+impl<'tcx> MovedVariablesCtxt<'tcx> {
+    fn get_required_kind(&self, place: &euv::Place<'tcx>, ref_hir_id: HirId) -> UpvarCapture {
+        if self
+            .moved
+            .iter()
+            .any(|upvar_ref| upvar_ref.0 == *place || upvar_ref.1 == ref_hir_id)
+        {
+            UpvarCapture::ByValue
+        } else {
+            self.captured
+                .iter()
+                .find(|upvar_ref| upvar_ref.1 == ref_hir_id)
+                .map_or(UpvarCapture::ByRef(ty::BorrowKind::ImmBorrow), |it| {
+                    UpvarCapture::ByRef(it.2)
+                })
+        }
+    }
+}
+
+impl<'tcx> euv::Delegate<'tcx> for MovedVariablesCtxt<'tcx> {
+    fn consume(&mut self, cmt: &euv::PlaceWithHirId<'tcx>, hir_id: HirId) {
+        if let euv::PlaceBase::Upvar(_) = cmt.place.base {
+            self.moved.push((cmt.place.clone(), hir_id));
```

Comment:
Is there any reason we can't store `&'tcx Place<'tcx>` to prevent cloning? This is cheap if projections is empty, tho you're cloning a `Vec` otherwise

Reply:
The `cmt` is not a `&'tcx` reference, so the lifetimes don't add up and the borrow checker is unhappy.